oci/spec_opts_unix.go

```diff
@@ -257,45 +257,45 @@ func WithUIDGID(uid, gid uint32) SpecOpts {
 // uid, and not returns error.
 func WithUserID(uid uint32) SpecOpts {
 	return func(ctx context.Context, client Client, c *containers.Container, s *specs.Spec) (err error) {
-		// TODO: support non-snapshot rootfs
-		if c.Snapshotter == "" {
-			return errors.Errorf("no snapshotter set for container")
-		}
-		if c.SnapshotKey == "" {
-			return errors.Errorf("rootfs snapshot not created for container")
+		if c.Snapshotter == "" && c.SnapshotKey == "" {
+			if s.Root.Path == "" {
+				if c.Snapshotter == "" {
```

Comment:
This `if` (and the following one) seem redundant because the previous line (260) already checked this condition. I think this could error immediately with "no shapshotter and rootfs set for container".

Or is 260 supposed to be `||` ? With `&&` it means that if one of them is the empty string the "else" block below could fail in an unexpected way.

Reply:
These errors are only returned when all 3 are empty.  And there are separate error messages for each missing field.  I think i retained that same logic with the change.

Reply:
The second missing field block (line 265) would never be hit, it's dead code, and there are unhandled error cases:

`c.Snapshotter` | `c.SnapshotKey` | `s.Root.Path` | result
---------------------|------------------------|--------------------|--------
"" | "" | "" | `errors.Errorf("no snapshotter set for container")`
"overlay" | "" | n/a | (goes to line 282 and may error because snapshotkey is "")
"" | "foo" | n/a | (goes to line 282 and may error because Snapshotter is "")
"overlay" | "foo" | n/a | (goes to line 282 and shouldn't error)
"" | "" | `/something` | (goes to line 269 and shouldn't error)

right?

Reply:
I think you are reading it wrong on your second line in the table. It won't go to 282, it goes to 265 and errors.

Reply:
same for the third line

Reply:
If it was `||` it would go to 265, but it's `&&` on 260.